sim_pkg/__main__.py

```diff
@@ -6,85 +8,37 @@
 import socketserver
 import csv
 import numpy as np
-from init_pose import init
 import argparse
+from sim_pkg.config import Configuration
 
-# sim_process_ = 0 
 robot_processes_ = []
-# vis_processes_ = 0
 parser = argparse.ArgumentParser(description='Run the simulator')
 
-parser.add_argument('-fn', '--filename', type=str, help="Name of the file without the .py extension")
+parser.add_argument('-fn', '--filename', type=str,
+                    help="Name of the file without the .py extension")
 
 args = parser.parse_args()
 
-def write_to_csv(free_port):
-    """
-    Write avaliable port number to csv file
-    """
-    f = open('port.csv', 'w')
-    num = [free_port]
-    with f:
-        # create the csv writer
-        writer = csv.writer(f)
-
-        # write a row to the csv file
-        writer.writerow(num)
-
-    # close the file
-    f.close()
-
-def init_pos(swarmsize):
-    """
-    Initialize the position
-    """
-    x = [0]*swarmsize
-    y = [0]*swarmsize
-    theta = [0]*swarmsize
-    a_ids = [0]*swarmsize
-
-    x, y, theta, a_ids = init(swarmsize,x,y,theta,a_ids)
-    id = [i for i in range(swarmsize)]
-
-    data = np.array([id, x, y, theta, a_ids]).T
-    np.savetxt("init.csv", data, delimiter=",", fmt="%.2f",
-           header="ID, x, y, theta, a_ids")
-
-
 def run():
     """
     Opens all the programs involved
     """
 
-    # Finds the first available socket on the system
-    with socketserver.TCPServer(("localhost", 0), None) as s:
-        free_port = s.server_address[1]
-
-    # Writes the socket number to port.csv file 
-    write_to_csv(free_port)
-    
     # Read required parameters from config.json
+    config = Configuration.from_path('config.json')  # Shouldn't be hardcoded.
     path = str(Path(__file__).parent)
-    with open('config.json', 'r') as myfile:
-        data=myfile.read()
-    config_var = json.loads(data)
-    num = config_var["NUMBER_OF_ROBOTS"]
-    use_vis = config_var["USE_VIS"]
-    # Initialize the positions in a csv file 
-    init_pos(num)
     
-    # print(path)
     # Open the simulator in a process
     sim_process_ = subprocess.Popen(['python3','-O', 'simulator.py'],close_fds=True,cwd=path)
```

Comment:
Why? I don't see the purpose of spawning a new process for this. I would much prefer this to be just a new thread?

I see tons of disadvantages to this method:
* You lose control of the process and can only communicate with it via kernel calls.
* You basically assume that `simulator.py` is just a valid program on someone's system -- distributing your software is a nightmare. I mean that software is usually distributed as packages and it's a nightmare to make a package out of your software. Usually in python we do a [traditional structure](https://packaging.python.org/en/latest/tutorials/packaging-projects/) that is easy to install with pip. With your structure this is difficult. Part of the problem is this `subprocess.Popen` calls you have everywhere.

I would much more prefer this to be a new thread or a new [multirprocessing.Process](https://docs.python.org/3/library/multiprocessing.html)

Reply:
Another big problem here is that it's a pain in the ass to communicate with your child processes. `visualizer` and `simulator` would not need to communicate over sockets if all you did is was use [multiprocessing.Pipe](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Pipe)